### PR TITLE
Add a system for explicitly marking const fns as promotable

### PR DESCRIPTION
**Pull request by oli-obk**
r? @nikomatsakis 

addresses the last point of https://gist.github.com/nikomatsakis/eee212d7c4a49ae75e19765c1e06bf49

### CODE REVIEW

**Comment by gnzlbg** [comment]
As mentioned on the `stdsimd` PR: the feature has no tracking issue, neither the PR nor the commits explain what it does, nor there are any docs... yet `std` library APIs  need to be marked with the attribute.

I don't know where the right place to document this is, but I think it should be explained somewhere (and maybe the libs team should be pinged here as well), since `std` maintainers need to know what this is and how to use it.

**Comment by oli-obk** [comment]
The idea is that you do not need to mark anything as promotable. We just needed an opt-in for preventing the automatic promotion of any const fn call due to https://github.com/rust-lang/rust/issues/50814 . Without that opt in, further stabilization of const fns is not something we want to do.

File: src/librustc_mir/transform/qualify_consts.rs [hunk 1]
```diff
@@ -929,10 +935,18 @@ This does not pose a problem by itself because they can't be accessed directly."
 
                             _ => {}
                         }
+                        // we can always mark intrinsics as promotable as they are inherently
+                        // unstable
+                        promotable = true;
```
Comment (nikomatsakis) [inline src/librustc_mir/transform/qualify_consts.rs]: let's extend the "promotable" whitelist to intrinsics too

**Review by nikomatsakis — COMMENTED** [review]
I approve of the general approach — but I agree that we should document. I was gonna say the same thing. @oli-obk and I talked and agreed we'd put it in the unstable reference.

**Comment by nikomatsakis** [comment]
cc @eddyb — do you want to take a look at this PR? (Not strictly necessary, I don't think) It adds a separate attribute to mark const fns as promotable.

---

@gnzlbg as I wrote above, I agree we need docs — but I think the short version is that we would only very rarely (if ever) want to add this attribute, and only after consulting with @oli-obk and/or myself. Basically we want to be conservative about what we will automatically promote to static lifetime (that is, you write `let x = &foo` and we allow that to escape the fn) until we've fully worked out the rules.

**Comment by oli-obk** [comment]
cc @eddyb

**Comment by eddyb** [comment]
I'm not sure we should mark any `const fn` that has *anything* in it other than creating ADTs.
There is a risk that *any* operation may fail when applied to an unknown value.

**Comment by pietroalbini** [comment]
Ping from triage @nikomatsakis! This PR needs your review.

**Comment by oli-obk** [comment]
closing in favour of https://github.com/rust-lang/rust/pull/51570

**Pull request closed without merging**